src/jit/SimdArm32Inl.h

```diff
@@ -523,14 +663,196 @@ static void emitBinarySIMD(sljit_compiler* compiler, Instruction* instr)
+    case ByteCode::I8X16ShrSOpcode:
+    case ByteCode::I8X16ShrUOpcode:
+        mask = SimdOp::shift8;
+        type = SLJIT_SIMD_ELEM_8;
+        break;
+    case ByteCode::I16X8ShlOpcode:
+        isShr = false;
+        FALLTHROUGH;
+    case ByteCode::I16X8ShrSOpcode:
+    case ByteCode::I16X8ShrUOpcode:
+        mask = SimdOp::shift16;
+        type = SLJIT_SIMD_ELEM_16;
+        break;
+    case ByteCode::I32X4ShlOpcode:
+        isShr = false;
+        FALLTHROUGH;
+    case ByteCode::I32X4ShrSOpcode:
+    case ByteCode::I32X4ShrUOpcode:
+        mask = SimdOp::shift32;
+        type = SLJIT_SIMD_ELEM_32;
+        break;
+    case ByteCode::I64X2ShlOpcode:
+        isShr = false;
+        FALLTHROUGH;
+    case ByteCode::I64X2ShrSOpcode:
+    case ByteCode::I64X2ShrUOpcode:
+        mask = SimdOp::shift64;
+        type = SLJIT_SIMD_ELEM_64;
+        break;
+    default:
+        ASSERT_NOT_REACHED();
+    }
+
+    simdOperandToArg(compiler, operands, args[0], type, SLJIT_FR0);
+    args[1].set(operands + 1);
+
+    args[2].set(operands + 2);
+    sljit_s32 dst = GET_TARGET_REG(args[2].arg, SLJIT_FR2);
+
+    if (SLJIT_IS_IMM(args[1].arg)) {
+        if (isShr) {
+            args[1].argw = (args[1].argw ^ mask) + 1;
+        }
+
+        args[1].argw &= mask;
+        if (args[1].argw == 0) {
+            if (args[2].arg != args[0].arg) {
+                sljit_emit_simd_mov(compiler, SLJIT_SIMD_STORE | SLJIT_SIMD_REG_128 | type, args[0].arg, args[2].arg, args[2].argw);
+            }
+            return;
+        }
+    } else {
+        sljit_s32 srcReg = GET_SOURCE_REG(args[1].arg, SLJIT_R0);
+        MOVE_TO_REG(compiler, SLJIT_MOV32, srcReg, args[1].arg, args[1].argw);
+
+        sljit_emit_op2(compiler, SLJIT_AND32, srcReg, 0, srcReg, 0, SLJIT_IMM, mask);
+        sljit_emit_simd_replicate(compiler, SLJIT_SIMD_REG_128 | SLJIT_SIMD_ELEM_8, dst, srcReg, 0);
+    }
+
+    switch (instr->opcode()) {
+    case ByteCode::I8X16ShlOpcode:
+        if (SLJIT_IS_IMM(args[1].arg)) {
+            simdEmitOp(compiler, SimdOp::Type::vshlImm | SimdOp::SHL8 | (args[1].argw << 16), dst, 0, args[0].arg);
+        } else {
+            simdEmitOp(compiler, SimdOp::Type::vshl | SimdOp::I8, dst, dst, args[0].arg);
+        }
+        break;
+    case ByteCode::I8X16ShrSOpcode:
+        if (SLJIT_IS_IMM(args[1].arg)) {
+            simdEmitOp(compiler, SimdOp::Type::vshrImm | SimdOp::SHL8 | (args[1].argw << 16), dst, 0, args[0].arg);
+        } else {
+            simdEmitOp(compiler, SimdOp::Type::vneg | SimdOp::S8, dst, 0, dst);
+            simdEmitOp(compiler, SimdOp::Type::vshl | SimdOp::I8, dst, dst, args[0].arg);
+        }
+        break;
+    case ByteCode::I8X16ShrUOpcode:
+        if (SLJIT_IS_IMM(args[1].arg)) {
+            simdEmitOp(compiler, SimdOp::Type::vshrImm | SimdOp::SHL8 | (args[1].argw << 16) | (0b1 << 24), dst, 0, args[0].arg);
+        } else {
+            simdEmitOp(compiler, SimdOp::Type::vneg | SimdOp::S8, dst, 0, dst);
```

Comment:
Would it be a good idea to do the neg before `sljit_emit_op2(compiler, SLJIT_AND32, srcReg, 0, srcReg, 0, SLJIT_IMM, mask);`